glean-core/src/metrics/timespan.rs

```diff
@@ -0,0 +1,106 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::sync::Mutex;
+use std::time::Instant;
+
+use crate::metrics::time_unit::TimeUnit;
+use crate::metrics::Metric;
+use crate::metrics::MetricType;
+use crate::storage::StorageManager;
+use crate::CommonMetricData;
+use crate::Glean;
+
+/// A timespan metric.
+///
+/// Timespans are used to make a measurement of how much time is spent in a particular task.
+#[derive(Debug)]
+pub struct TimespanMetric {
+    meta: CommonMetricData,
+    time_unit: TimeUnit,
+    start_time: Mutex<Option<Instant>>,
+}
+
+impl MetricType for TimespanMetric {
+    fn meta(&self) -> &CommonMetricData {
+        &self.meta
+    }
+
+    fn meta_mut(&mut self) -> &mut CommonMetricData {
+        &mut self.meta
+    }
+}
+
+impl TimespanMetric {
+    /// Create a new timespan metric.
+    pub fn new(meta: CommonMetricData, time_unit: TimeUnit) -> Self {
+        Self {
+            meta,
+            time_unit,
+            start_time: Mutex::new(None),
+        }
+    }
+
+    /// Start tracking time for the provided metric.
+    pub fn start(&self, glean: &Glean) {
+        if !self.should_record(glean) {
+            return;
+        }
+
+        let mut start_time = self.start_time.lock().unwrap();
```

Comment:
Since this code is run off of the main thread in a dispatch queue, this could happen significantly after the "start" of the thing being measured.  Even though it's less clean, it's probably better to capture the current time on the main thread on the Kotlin side and then pass it in over FFI.

Reply:
Hm, what about _not_ dispatching this? It doesn't hit the database here, so I/O should not be a problem.

(Unless ... we record an error, then it _does_ hit the database :/)

Reply:
Ah, but then we have the same problem on stop. Hmpf.

Reply:
I agree with @mdboom , good catch. Let's keep the dispatcher (at least for now) and propagate the time over FFI.

Reply:
Ok, fixed that now. Now missing double-checking that I have all the tests and same behavior everywhere.